File: Backend/models/calendar.py
"""
Calendar Models

Defines models for the Calendar feature:
- CustomReminder: User-created calendar reminders
- CalendarEvent: Dataclass for unified calendar event representation
- Enums for event types, status, priority
"""
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Optional, List, Dict, Any, TYPE_CHECKING
from enum import Enum
from uuid import UUID as PythonUUID, uuid4

import sqlalchemy as sa
from sqlalchemy import Column, DateTime
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlmodel import SQLModel, Field, Relationship


# ============================================================================
# ENUMS
# ============================================================================

class CalendarEventType(str, Enum):
    """Types of calendar events from various sources"""
    RENT_DUE = "rent_due"
    INVOICE_DUE = "invoice_due"
    LEASE_START = "lease_start"
    LEASE_EXPIRING = "lease_expiring"
    MAINTENANCE_SCHEDULED = "maintenance_scheduled"
    INSURANCE_EXPIRY = "insurance_expiry"
    MORTGAGE_RENEWAL = "mortgage_renewal"
    CUSTOM_REMINDER = "custom_reminder"


class CalendarEventStatus(str, Enum):
    """Computed status based on dates and completion"""
    UPCOMING = "upcoming"  # Future event
    DUE = "due"  # Due today
    OVERDUE = "overdue"  # Past due and not completed
    COMPLETED = "completed"  # Resolved/completed


class CalendarEventPriority(str, Enum):
    """Event priority level"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


# ============================================================================
# DATABASE MODEL
# ============================================================================

class CustomReminder(SQLModel, table=True):
    """
    User-created calendar reminders.
    
    This is the only calendar-specific table - all other events are
    computed from existing source tables (invoices, leases, etc.)
    """
    __tablename__ = "custom_reminders"
    
    # Primary key
    id: PythonUUID | None = Field(
        default=None,
        sa_column=Column(
            PG_UUID(as_uuid=True),
            primary_key=True,
            server_default=sa.text("gen_random_uuid()")
        )
    )
    
    # Owner
    user_id: PythonUUID = Field(foreign_key="users.id")
    
    # Content
    title: str = Field(max_length=255)
    description: Optional[str] = Field(default=None)
    
    # Temporal
    reminder_date: datetime = Field(
        sa_column=Column(DateTime(timezone=True), nullable=False, index=True)
    )
    all_day: bool = Field(default=False)
    
    # Optional associations
    property_id: Optional[int] = Field(default=None, foreign_key="properties.id")
    unit_id: Optional[int] = Field(
        default=None,
        foreign_key="property_units.id"
    )
    tenant_id: Optional[int] = Field(
        default=None,
        foreign_key="tenants.id"
    )
    
    # Status
    is_completed: bool = Field(default=False, index=True)
    completed_at: Optional[datetime] = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=True)
    )
    
    # Notifications
    notify_before_hours: int = Field(default=24)
    notified_at: Optional[datetime] = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=True),
        description="Timestamp when notification was sent (prevents duplicates)"
    )
    notification_id: Optional[PythonUUID] = Field(
        default=None,
        foreign_key="notifications.id",
        description="Reference to created notification record"
    )
    
    # Timestamps
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        sa_column=Column(DateTime(timezone=True), nullable=False)
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc),
        sa_column=Column(DateTime(timezone=True), nullable=False)
    )
    
    # Relationships
    if TYPE_CHECKING:
        from Backend.models.user import User
        from Backend.models.property import Property
        from Backend.models.units import PropertyUnit
        from Backend.models.tenant import Tenant
        from Backend.models.notification import Notification
    
    user: Optional["User"] = Relationship(back_populates="custom_reminders")
    property: Optional["Property"] = Relationship()
    unit: Optional["PropertyUnit"] = Relationship()
    tenant: Optional["Tenant"] = Relationship()
    notification: Optional["Notification"] = Relationship()


# ============================================================================
# DATACLASS (NOT A DATABASE TABLE)
# ============================================================================

@dataclass
class CalendarEvent:
    """
    Unified calendar event representation.
    
    This is NOT a database table - it's a runtime dataclass built from
    various source tables (invoices, leases, maintenance, custom_reminders).
    
    Each event has a composite ID like "invoice_123" or "lease_start_456"
    to identify the source.
    """
    # Identity
    id: str  # Format: "{source_type}_{source_id}" e.g., "invoice_123"
    type: CalendarEventType
    
    # Content
    title: str
    description: Optional[str]
    
    # Temporal
    start_at: datetime
    end_at: Optional[datetime]
    all_day: bool
    
    # Status (computed at runtime)
    status: CalendarEventStatus
    priority: CalendarEventPriority
    
    # Associations (nullable)
    property_id: Optional[int]
    property_name: Optional[str]  # Denormalized for display
    unit_id: Optional[int]
    tenant_id: Optional[int]
    tenant_name: Optional[str]  # Denormalized for display
    lease_id: Optional[int]
    
    # Source tracking
    source_type: str  # "invoice", "lease", "maintenance", "property", "custom"
    source_id: int | str  # ID in the source table
    
    # UI helpers (computed)
    color: str  # "green", "amber", "red" based on status
    quick_actions: List[str]  # e.g., ["send_invoice", "record_payment"]
    
    # Metadata (flexible JSON for type-specific data)
    metadata: Dict[str, Any] = field(default_factory=dict)

File: Frontend/src/utils/api/calendar.ts
/**
 * Calendar API Client
 * 
 * Handles all API calls for the calendar feature including:
 * - Fetching unified calendar events
 * - Creating/updating/deleting custom reminders
 */

import { apiRequest, formatQueryString } from './core';

// ============================================================================
// TYPES
// ============================================================================

export interface CalendarFilters {
  from_date: string;
  to_date: string;
  property_id?: number;
  unit_id?: number;
  tenant_id?: number;
  event_type?: CalendarEventType;
  status?: CalendarEventStatus;
}

export enum CalendarEventType {
  INVOICE_DUE = 'invoice_due',
  LEASE_START = 'lease_start',
  LEASE_EXPIRING = 'lease_expiring',
  MAINTENANCE_SCHEDULED = 'maintenance_scheduled',
  INSURANCE_EXPIRY = 'insurance_expiry',
  MORTGAGE_RENEWAL = 'mortgage_renewal',
  CUSTOM_REMINDER = 'custom_reminder',
}

export enum CalendarEventStatus {
  UPCOMING = 'upcoming',
  DUE = 'due',
  OVERDUE = 'overdue',
  COMPLETED = 'completed',
}

export enum CalendarEventPriority {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
}

export interface RelatedEntity {
  id: number;
  name: string;
}

export interface CalendarEvent {
  id: string;
  type: CalendarEventType;
  title: string;
  description?: string;
  start_at: string;
  end_at?: string;
  all_day: boolean;
  status: CalendarEventStatus;
  priority: CalendarEventPriority;
  link?: string;
  property?: RelatedEntity;
  property_id?: number;
  property_name?: string;
  unit?: RelatedEntity;
  unit_id?: number;
  tenant?: RelatedEntity;
  tenant_id?: number;
  tenant_name?: string;
  lease_id?: number;
  source_type?: string;
  source_id: string;
  color: string;
  quick_actions: string[];
  metadata: Record<string, any>;
}

export interface CalendarEventsResponse {
  events: CalendarEvent[];
  total: number;
  from_date: string;
  to_date: string;
}

export interface CustomReminderCreate {
  title: string;
  description?: string;
  reminder_date: string;
  all_day?: boolean;
  property_id?: number;
  unit_id?: number;
  tenant_id?: number;
  notify_before_hours?: number;
}

export interface CustomReminderUpdate {
  title?: string;
  description?: string;
  reminder_date?: string;
  all_day?: boolean;
  is_completed?: boolean;
  property_id?: number;
  unit_id?: number;
  tenant_id?: number;
  notify_before_hours?: number;
}

export interface CustomReminder {
  id: string;
  user_id: string;
  title: string;
  description?: string;
  reminder_date: string;
  all_day: boolean;
  property_id?: number;
  unit_id?: number;
  tenant_id?: number;
  is_completed: boolean;
  completed_at?: string;
  notify_before_hours: number;
  created_at: string;
  updated_at: string;
}

// ============================================================================
// API FUNCTIONS
// ============================================================================

/**
 * Fetch unified calendar events from all sources
 */
export const fetchCalendarEvents = async (
  filters: CalendarFilters
): Promise<CalendarEventsResponse> => {
  const params = new URLSearchParams();
  
  if (filters.from_date) params.append('from_date', filters.from_date);
  if (filters.to_date) params.append('to_date', filters.to_date);
  if (filters.property_id) params.append('property_id', filters.property_id.toString());
  if (filters.unit_id) params.append('unit_id', filters.unit_id.toString());
  if (filters.tenant_id) params.append('tenant_id', filters.tenant_id.toString());
  if (filters.event_type) params.append('event_type', filters.event_type);
  if (filters.status) params.append('status', filters.status);

  const queryString = params.toString();
  return apiRequest(`/calendar/events${formatQueryString(queryString)}`);
};

/**
 * Create a new custom reminder
 */
export const createCustomReminder = async (
  reminderData: CustomReminderCreate
): Promise<CustomReminder> => {
  return apiRequest('/calendar/reminders', {
    method: 'POST',
    body: JSON.stringify(reminderData),
  });
};

/**
 * Update an existing custom reminder
 */
export const updateCustomReminder = async (
  reminderId: string,
  updateData: CustomReminderUpdate
): Promise<CustomReminder> => {
  return apiRequest(`/calendar/reminders/${reminderId}`, {
    method: 'PATCH',
    body: JSON.stringify(updateData),
  });
};

/**
 * Delete a custom reminder
 */
export const deleteCustomReminder = async (reminderId: string): Promise<void> => {
  return apiRequest(`/calendar/reminders/${reminderId}`, {
    method: 'DELETE',
  });
};

/**
 * Mark a custom reminder as complete
 */
export const completeReminder = async (reminderId: string): Promise<CustomReminder> => {
  return updateCustomReminder(reminderId, { is_completed: true });
};

/**
 * Get a specific custom reminder
 */
export const getCustomReminder = async (reminderId: string): Promise<CustomReminder> => {
  return apiRequest(`/calendar/reminders/${reminderId}`);
};

